src/metabase/models/serialization.clj

```diff
@@ -210,16 +210,42 @@
   [s]
   (boolean (re-matches #"^[0-9a-fA-F]{8}$" s)))
 
+;; ## Memoizing `hydrated-hash`
+;;
+;; Hashing a Field requires its Table; hashing a Table requires its Database.
+;; Letting each of those hit the appdb for every Field lookup (when it lacks an `entity_id`) is too costly,
+;; so we cache any that have to be looked up right here.
+;;
+;; Memory use is not a serious concern here, for two reasons:
+;; 1. This is caching the `hydrated-hash` lookups, so it doesn't cache Fields but only Tables and Databases.
+;; 2. This is called only when [[backfill-entity-id]] needs to generate an `entity_id` by hashing. Once the background
+;;    job populates that column everywhere, this will always be empty.
+;;
+;; NOTE: To support Metabase upgrades where the new `entity_id`s might still be blank, this code will have to live on.
+;; But in practice once `entity_id`s are populated this cache will never be needed.
+(def ^:private hydrated-hash-cache
+  (atom {}))
```

Comment:
Do we have a sense for what the worst case scenario would be here, memory consumption wise? I don't have a great sense of how large our largest instances are. If someone has a ridiculous number of tables and is using the new cache prewarming feature or whatever. How quickly / often will the backfill task run, etc.

Reply:
The largest instances we know about have O(10M) fields, so 1M tables works as a pessimistic estimate. Even if all of them were accessed and the backfill never ran, this cache would be holding a map with 1M entries, which are a `long` and a 21-char `String` each, plus a handful of pointers' worth of overhead. probably at least 2 or 3 pointers of overhead. So perhaps `8 + 64 + 8*4 = 104` bytes each? That's O(100MB) total, which while nontrivial is not going to knock over a big instance with 10M+ fields.